File: vmlib/src/main/java/me/shouheng/vmlib/base/BaseService.java
package me.shouheng.vmlib.base;

import android.app.Service;
import me.shouheng.vmlib.anno.ServiceConfiguration;
import me.shouheng.vmlib.bus.Bus;

/**
 * Base service to handle {@link ServiceConfiguration} annotation.
 *
 * @author <a href="mailto:shouheng2015@gmail.com">WngShhng</a>
 * @version 2019-7-1
 */
public abstract class BaseService extends Service {

    private boolean useEventBus = false;

    public BaseService() {
        super();
        ServiceConfiguration configuration = this.getClass().getAnnotation(ServiceConfiguration.class);
        if (configuration != null) {
            useEventBus = configuration.useEventBus();
        }
    }

    @Override
    public void onCreate() {
        super.onCreate();
        if (useEventBus) {
            Bus.get().register(this);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (useEventBus) {
            Bus.get().unregister(this);
        }
    }
}

File: vmlib/src/main/java/me/shouheng/vmlib/base/BaseService.kt
package me.shouheng.vmlib.base

import android.app.Service
import androidx.lifecycle.LifecycleOwner
import me.shouheng.vmlib.bus.Bus
import me.shouheng.vmlib.anno.ServiceConfiguration
import me.shouheng.vmlib.component.SimpleLifecycle

/**
 * Base service to handle [ServiceConfiguration] annotation.
 *
 * @author [ShouhengWang](mailto:shouheng2020@gmail.com)
 * @version 2019-7-1
 */
abstract class BaseService : Service(), LifecycleOwner {

    private var useEventBus = false

    private val lifecycle = SimpleLifecycle(this)

    override fun onCreate() {
        super.onCreate()
        if (useEventBus) {
            Bus.get().register(this)
        }
        lifecycle.onCreate()
    }

    override fun onDestroy() {
        super.onDestroy()
        if (useEventBus) {
            Bus.get().unregister(this)
        }
        lifecycle.onDestroy()
    }

    override fun getLifecycle() = lifecycle

    init {
        val configuration = this.javaClass.getAnnotation(
            ServiceConfiguration::class.java
        )
        if (configuration != null) {
            useEventBus = configuration.useEventBus
        }
    }
}